clis/teliod/src/qnap.rs

```diff
@@ -0,0 +1,398 @@
+const PID_FILE: &str = concatcp!(TELIOD_TMP_DIR, "/teliod.pid");
+const QPKG_DIR: &str = "/share/CACHEDEV1_DATA/.qpkg/NordSecurityMeshnet";
+const TELIOD_BIN: &str = concatcp!(QPKG_DIR, "/teliod");
+const MESHNET_LOG: &str = concatcp!(QPKG_DIR, "/meshnet.log");
+const TELIOD_LOG: &str = "/var/log/teliod.log";
+
+#[cfg(not(test))]
+const TELIOD_CFG: &str = concatcp!(QPKG_DIR, "/teliod.cfg");
+#[cfg(test)]
+use tests::TELIOD_CFG;
+
+pub(crate) fn handle_request(request: Request) -> Response {
+    match (request.method(), request.uri().query()) {
+        (&Method::POST, _) => start_daemon(),
+        (&Method::DELETE, _) => stop_daemon(),
+        (&Method::PATCH, _) => {
+            let body = match String::from_utf8(request.into_body()) {
+                Ok(body) => body,
+                Err(_) => return text_response(400, "Invalid UTF-8 in request body."),
+            };
+            update_config(&body)
+        }
+        (&Method::GET, Some("info=get-status")) => get_status(),
+        (&Method::GET, Some("info=get-teliod-logs")) => get_teliod_logs(),
+        (&Method::GET, Some("info=get-meshnet-logs")) => get_meshnet_logs(),
+        (_, _) => text_response(400, "Invalid request."),
+    }
+}
+
+fn is_teliod_running() -> bool {
+    fs::read_to_string(PID_FILE)
+        .and_then(|pid| fs::metadata(format!("/proc/{}", pid.trim())))
+        .is_ok()
+}
+
+fn teliod_socket_exists() -> bool {
+    DaemonSocket::get_ipc_socket_path()
+        .and_then(|path| path.metadata())
+        .is_ok()
+}
+
+fn kill_teliod_process() -> Result<(), io::Error> {
+    match fs::read_to_string(PID_FILE) {
+        Ok(pid) => {
+            let _ = Command::new("kill").arg("-9").arg(pid.trim()).status();
+            if teliod_socket_exists() {
+                let _ = DaemonSocket::get_ipc_socket_path().and_then(fs::remove_file);
+            }
+            fs::remove_file(PID_FILE)
+        }
+        Err(e) => Err(e),
+    }
+}
+
+fn start_daemon() -> Response {
+    if is_teliod_running() {
+        if teliod_socket_exists() {
+            return text_response(400, "Application is already running.");
+        } else {
+            let _ = kill_teliod_process();
+        }
+    }
+
+    let teliod_log_file = match fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(TELIOD_LOG)
+    {
+        Ok(file) => file,
+        Err(_) => {
+            return text_response(500, "Failed to open teliod log file.");
+        }
+    };
+
+    match Command::new("setsid")
+        .arg(TELIOD_BIN)
+        .arg("daemon")
+        .arg(TELIOD_CFG)
+        .stdout(Stdio::from(
+            teliod_log_file
+                .try_clone()
+                .expect("Failed to open teliod log file."),
+        ))
+        .stderr(Stdio::from(teliod_log_file))
+        .spawn()
+    {
+        Ok(child) => {
+            let pid = child.id();
+            if fs::write(PID_FILE, pid.to_string()).is_ok() {
```

Comment:
This should be done by the deamon process, not by the cgi process.

Reply:
This is a QNAP specific file, I can do it conditionally when the `qnap` feature is enabled. wdyt?

Reply:
Ok :)